_posts/2025-12-01-idempotency-in-apis.md

```diff
@@ -0,0 +1,145 @@
+Does it mean *"delete entity with ID={id}"*,
+or does it mean *"make sure entity with ID={id} is non-existent"*?
+An entity can be deleted only once, but when the actual delete action becomes an implementation detail,
+the API can return `200 OK` regardless of whether the entity was just deleted,
+previously deleted, or maybe never existed.
+This behavior is often precisely what you need and leads to a more forgiving API,
+allowing simpler, more robust code for the API consumer.
+
+As we move to the *Create* operation with `POST` requests,
+it gets even murkier.
+Consider the following `POST /car` request body:
+
+```json
+{
+    "model": "Nimbus",
+    "year": 2022
+}
+```
+
+Two `POST` requests will result in two entities being created.
+The API server enables this
+by automatically assigning each entity an ID before storing it (e.g., in a database).
+
+There are a couple of ways to deal with this.
+One way is to leverage the [`Idempotency-Key` header][2]
+containing a unique ID for a single request across multiple retries.
+But this is no magic trick that will fix everything
+— it still requires the implementation to respect the header.
+In other words: It is still the API provider's responsibility
+to limit multiple identical requests to a single execution.
+
+The other way is slightly less explicit about it,
+but has served me well on many occasions.
+Instead of having the server auto-generate IDs for entities,
+I force the caller to provide an ID for the entity:
+
+```json
+{
+    "id": "79b72ba6-939b-40f7-a8e9-44c3d52da926",
+    "model": "Nimbus",
+    "year": 2022
+}
+```
+
+Again, just like with *Delete*, it is about the interpretation of the operation.
+Does it mean *"create entity with the ID={id}"*,
+or does it mean *"make sure entity with the ID={id} exists"*.
+The latter can, in good conscience,
+return `200 OK` even though it didn't create the entity by itself.
+
+What I like about this approach is,
+that it is impossible to opt out of using the API in a way that doesn't support retries.
+The consumer doesn't need to understand or even see the word "idempotency".
+
+Many storage systems make this behavior straightforward to implement.
+The Postgres database for instance, has an [`ON CONFLICT` clause][3]
+to specify an alternative action when the ID already exists.
+Instead of raising an error or creating a duplicate, the database action can be `DO NOTHING`.
+
+Lastly, the *Update* operation using a `PATCH /car/{id}` request
+is similar to the *Create* operation in many ways.
+But it takes slightly more effort to avoid introducing non-idempotent behavior.
+I.e., auto-incrementing a counter would break idempotency.
+Idempotency could still be achieved by using the `Idempotency-Key` header instead,
+but the implementation would become more complex.
+
+Remember to include only the properties that you want to update in a `PATCH` request.
+If you want to overwrite the entire thing, you are probably looking for the `PUT` request.
```

Comment:
These two could use a "Note" or "Tip" kind of admonition/callout for the newbies out there. If you don't have such styling or components available, a simple bold paragraph with Note or Tip above these two sentences will do the trick.